Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean

```diff
@@ -0,0 +1,201 @@
+  DFunLike.ext _ _ <| by simp [ZMod.intCast_surjective.forall, zmod]
+
+variable {n}
+
+@[simp] lemma zmod_add : ∀ x y : ZMod n, zmod n (x + y) = zmod n x * zmod n y := by
+  simp [DFunLike.ext_iff, ← Int.cast_add, zmod, add_mul, add_div, map_add_eq_mul]
+
+lemma zmod_injective : Injective (zmod n) := by
+  simp_rw [Injective, ZMod.intCast_surjective.forall]
+  rintro x y h
+  replace hn : (n : ℝ) ≠ 0 := by have := NeZero.ne n; positivity
+  simpa [pi_ne_zero, exp_inj, hn, CharP.intCast_eq_intCast (ZMod n) n] using
+    (zmod_intCast ..).symm.trans <| (DFunLike.congr_fun h ((1 : ℤ) : ZMod n)).trans <|
+      zmod_intCast ..
+
+@[simp] lemma zmod_inj {x y : ZMod n} : zmod n x = zmod n y ↔ x = y := zmod_injective.eq_iff
+
+/-- `AddChar.zmod` bundled as an `AddChar`. -/
+def zmodHom : AddChar (ZMod n) (AddChar (ZMod n) Circle) where
+  toFun := zmod n
+  map_zero_eq_one' := by simp
+  map_add_eq_mul' := by simp
+
+/-- Character on a product of `ZMod`s given by `x ↦ ∏ i, e ^ (2 * π * I * x i * y / n)`. -/
+private def mkZModAux {ι : Type} [DecidableEq ι] (n : ι → ℕ) [∀ i, NeZero (n i)]
+    (u : ∀ i, ZMod (n i)) : AddChar (⨁ i, ZMod (n i)) Circle :=
+  AddChar.directSum fun i ↦ zmod (n i) (u i)
+
+private lemma mkZModAux_injective {ι : Type} [DecidableEq ι] {n : ι → ℕ} [∀ i, NeZero (n i)] :
+    Injective (mkZModAux n) :=
+  AddChar.directSum_injective.comp fun f g h ↦ by simpa [funext_iff] using h
+
+/-- The circle-valued characters of a finite abelian group are the same as its complex-valued
+characters. -/
+def circleEquivComplex [Finite α] : AddChar α Circle ≃+ AddChar α ℂ where
+  toFun ψ := toMonoidHomEquiv.symm <| coeHom.comp ψ.toMonoidHom
+  invFun ψ :=
+    { toFun := fun a ↦ (⟨ψ a, mem_sphere_zero_iff_norm.2 <| ψ.norm_apply _⟩ : Circle)
+      map_zero_eq_one' := by simp [Circle]
+      map_add_eq_mul' := fun a b ↦ by ext : 1; simp [map_add_eq_mul] }
+  left_inv ψ := by ext : 1; simp
+  right_inv ψ := by ext : 1; simp
+  map_add' ψ χ := rfl
+
+@[simp] lemma card_eq [Fintype α] : card (AddChar α ℂ) = card α := by
+  obtain ⟨ι, _, n, hn, ⟨e⟩⟩ := AddCommGroup.equiv_directSum_zmod_of_finite' α
+  classical
+  have hn' i : NeZero (n i) := by have := hn i; exact ⟨by positivity⟩
+  let f : α → AddChar α ℂ := fun a ↦ coeHom.compAddChar ((mkZModAux n <| e a).compAddMonoidHom e)
+  have hf : Injective f := circleEquivComplex.injective.comp
+    ((compAddMonoidHom_injective_left _ e.surjective).comp <| mkZModAux_injective.comp <|
+      DFunLike.coe_injective.comp <| e.injective.comp Additive.ofMul.injective)
+  exact (card_addChar_le _ _).antisymm (Fintype.card_le_of_injective _ hf)
+
+/-- `ZMod n` is (noncanonically) isomorphic to its group of characters. -/
+def zmodAddEquiv : ZMod n ≃+ AddChar (ZMod n) ℂ := by
+  refine AddEquiv.ofBijective
+    (circleEquivComplex.toAddMonoidHom.comp <| AddChar.toAddMonoidHom zmodHom) ?_
+  rw [Fintype.bijective_iff_injective_and_card, card_eq]
+  exact ⟨circleEquivComplex.injective.comp zmod_injective, rfl⟩
```

Comment:
One could then golf this to
```lean
def zmodAddEquiv : ZMod n ≃+ AddChar (ZMod n) ℂ :=
  addEquiv_self (ZMod n) |>.some.symm
```
but I assume you what the specific isomorphism here.

Reply:
Yes, the actual isomorphism here was definitely used in the original application of this code about a year and a half ago; I'm pretty sure it's useful in the application in LeanAPAP too.

Reply:
I actually don't think we ever use the specific isomorphism, no